ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Forge.hs

```diff
@@ -119,15 +145,35 @@ forgeBlock (WithEBBNodeConfig cfg) curSlot curNo prevHash txs isLeader = do
       , pbftProtocolMagic
       } = encNodeConfigExt cfg
 
+    blockPayloads :: BlockPayloads
+    blockPayloads = foldl' extendBlockPayloads initBlockPayloads txs
+
     txPayload :: CC.UTxO.TxPayload
-    txPayload = CC.UTxO.mkTxPayload (map (void . byronTx) txs)
+    txPayload = CC.UTxO.mkTxPayload (bpTxs blockPayloads)
+
+    dlgPayload :: CC.Delegation.Payload
+    dlgPayload = CC.Delegation.unsafePayload (bpDlgCerts blockPayloads)
+
+    updatePayload :: CC.Update.Payload
+    updatePayload = CC.Update.payload (bpUpProposal blockPayloads)
+                                      (bpUpVotes blockPayloads)
+
+    extendBlockPayloads :: BlockPayloads
+                        -> GenTx (ByronBlockOrEBB cfg)
+                        -> BlockPayloads
+    extendBlockPayloads bp@BlockPayloads{bpTxs, bpDlgCerts, bpUpVotes} genTx =
+      case genTx of
+        ByronTx             _ tx   -> bp { bpTxs        = void tx : bpTxs }
+        ByronDlg            _ cert -> bp { bpDlgCerts   = void cert : bpDlgCerts }
+        ByronUpdateProposal _ prop -> bp { bpUpProposal = Just (void prop) }
+        ByronUpdateVote     _ vote -> bp { bpUpVotes    = void vote : bpUpVotes }
 
```

Comment:
Hmm, it's a bit inefficient to throw all the serialisation here. I see there is `recoverProof` as well, can we use that?

Reply:
We can use `recoverProof`, we need functions to recover the `dlgPayload` and the `updatePayload`.

Unlike what I discussed in https://github.com/input-output-hk/ouroboros-network/issues/905#issuecomment-540076435, using `recoverProof` here would only be an optimisation, it is not essential. Can you add a TODO stating that we should try to use `recoverProof` for this in the future?

Reply:
> Can you add a TODO stating that we should try to use `recoverProof` for this in the future?

Put it in the body of `extendBlockPayloads`.